### PR TITLE
Add macros translator

### PR DESCRIPTION
**Pull request by smousy**

### CODE REVIEW

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

File: MacrosTranslator/src/translator/macros_translator.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+bool MacrosTranslator::Run(std::string const & workDirectoryPath, std::string const & targetDirectoryPath)
+{
+  ScDirectory const & workDirectory{workDirectoryPath};
+  m_filesCount = workDirectory.CountFiles(m_extensions);
+  ScDirectory const & targetDirectory{targetDirectoryPath};
+
+  std::string const & startDirectoryPath = workDirectory.GetPath();
+  m_fileNumber = 0;
+
+  std::cout << "SCn-tex sources directory: " << workDirectory.GetPath() << std::endl;
```
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/src/translator/macros_translator.cpp:12]: Done

File: MacrosTranslator/src/translator/commands/commonCommands/tex_unique_command.cpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    {
+        result = "\\begin{scnset}\n";
+    }
+    else if(command == "scnendstruct")
+    {
+        result = "\\end{scnstruct}\n";
+    }
+    else if(command == "scnstructinclusion")
+    {
+        result = "\\end{subset}{"+params.at(1)+"\n";
```
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/src/translator/commands/commonCommands/tex_unique_command.cpp:32]: Done

File: MacrosTranslator/src/translator/commands/commonCommands/tex_one_to_environment_command.cpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        end = content.find(delim, start);
+        items.push_back(content.substr(start, end - start));
+    }
+
+    std::string result = "\\begin{"+params.at(0)+"}\n";
+    for(const std::string& item:items)
+    {
+        result+="\\scnitem{"+item+"}\n";
+    }
```
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/src/translator/commands/commonCommands/tex_one_to_environment_command.cpp:22]: Done

File: MacrosTranslator/src/translator/commands/commonCommands/tex_command.cpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#include "tex_command.h"
+
+std::string TexCommand::Complete(
+        TexCommandParams const & params)
+{
+  std::string result = "\\"+params.at(0);
+  for(const std::string& param : params)
+  {
+      if(param!=params.at(0))
```
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/src/translator/commands/commonCommands/tex_command.cpp:9]: Done

File: MacrosTranslator/src/translator/commands/commonCommands/tex_command.cpp [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#include "tex_command.h"
+
+std::string TexCommand::Complete(
+        TexCommandParams const & params)
+{
+  std::string result = "\\"+params.at(0);
```
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/src/translator/commands/commonCommands/tex_command.cpp:6]: Done

File: MacrosTranslator/src/translator/commands/commands.h [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+#include "command_defines.h"
+
+using ScSCnTexCommands = std::map<std::set<std::string>, ScSCnTexCommand*>;
+
+static std::set<std::string> const latexCommandsname =
+        {
+        //if command begin with scn then this is not a latex command
+        };
+
+static std::set<std::string> const oneSameCommandsname =
```
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/src/translator/commands/commands.h:15]: Done

File: MacrosTranslator/src/translator/commands/commands.h [hunk 7]
```diff
@@ -0,0 +1,161 @@
+#pragma once
+
+#include <map>
+#include <set>
+
+#include "command_defines.h"
+
+using ScSCnTexCommands = std::map<std::set<std::string>, ScSCnTexCommand*>;
+
+static std::set<std::string> const latexCommandsname =
```
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/src/translator/commands/commands.h:10]: Done

File: MacrosTranslator/README.md [hunk 8]
```diff
@@ -0,0 +1,17 @@
+# Macros-translator
+
+Tex2SCs-translator is a tool for translating SCn-represented *.tex files into new macros *.tex files.
```
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/README.md:3]: Done

File: MacrosTranslator/README.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# Macros-translator
```
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/README.md:1]: Done

File: MacrosTranslator/src/translator/commands/commonCommands/tex_command.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/src/translator/commands/commonCommands/tex_command.cpp:9]: Done

File: MacrosTranslator/src/translator/commands/commonCommands/tex_unique_command.cpp [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+            end = content.find(delim, start);
+            items.push_back(content.substr(start, end - start));
+        }
+
+        result = "\\begin{scnhaselementrolelist}\n";
+        for(const std::string& item:items)
+        {
+            result+="\\scnitem{"+item+"}\n";
+        }
+        result+="\\end{scnhaselementrolelist}\n";
```
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/src/translator/commands/commonCommands/tex_unique_command.cpp:141]: Done

File: MacrosTranslator/src/translator/commands/commonCommands/tex_command.cpp [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+  for(const std::string& param : params)
+  {
+      if(param!=params.at(0))
+      {
+          result += "{" + param + "}";
```
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/src/translator/commands/commonCommands/tex_command.cpp:11]: Done

File: MacrosTranslator/src/translator/commands/commonCommands/tex_unique_command.cpp [hunk 12]
```diff
@@ -0,0 +1,161 @@
+#include "tex_unique_command.h"
+
+std::string TexUniqueCommand::Complete(
+        TexCommandParams const & params)
+{
+    const std::string& command = params.at(0);
+    std::string delim = ";";
+    std::string result;
+
+    if(command == "scnstartsubstruct")
```
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/src/translator/commands/commonCommands/tex_unique_command.cpp]: Done

File: MacrosTranslator/src/translator/commands/commonCommands/tex_unique_command.cpp [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+        }
+
+        result = "\\begin{scnhaselementrolelist}\n";
+        for(const std::string& item:items)
+        {
+            result+="\\scnitem{"+item+"}\n";
+        }
+        result+="\\end{scnhaselementrolelist}\n";
+    }
+    else if (command == "scnsdrelation")
```
[the remark this answers is not in the document]
Comment (smousy) [reply MacrosTranslator/src/translator/commands/commonCommands/tex_unique_command.cpp]: Done

File: MacrosTranslator/src/translator/commands/commonCommands/unique_commands.h [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+    };
+};
+
+class BeginCommand : public ScSCnTexCommand
+{
+public:
+    std::string Complete(TexCommandParams const &params) override
+    {
+        std::stringstream result;
+        result << texSlash << "begin" << openBr << GetFirstParameter(params) << closeBr << newline;
```
Comment (NikitaZotov) [inline MacrosTranslator/src/translator/commands/commonCommands/unique_commands.h:30]: let's name constants as "MY_CONSTANT"

File: MacrosTranslator/src/translator/commands/commonCommands/unique_commands.h [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+    };
+};
+
+class ScnSdClassCommand : public ScSCnTexCommand
+{
+public:
+    std::string Complete(TexCommandParams const &params) override
+    {
+        std::stringstream result;
+        result << texSlash << "begin" << openBr << "scnhaselementrolelist" << closeBr << openBr
```
Comment (NikitaZotov) [inline MacrosTranslator/src/translator/commands/commonCommands/unique_commands.h:263]: So, such code is hard anyway. I propose to create template functions to customize styles of strings

File: MacrosTranslator/src/translator/macros_translator.cpp [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    antlr4::CommonTokenStream tokens(&lexer);
+    SCnTexParser parser(&tokens);
+    SCnTexParser::ScnTexTextContext *ctx;
+    try
+    {
+        ctx = parser.scnTexText();
+    }
+    catch (...)
+    {
+        throw;
```
Comment (NikitaZotov) [inline MacrosTranslator/src/translator/macros_translator.cpp:77]: May be to throw exception that was caught?

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

File: MacrosTranslator/src/translator/macros_translator.cpp (hunk 16 above)
Comment (smousy) [reply MacrosTranslator/src/translator/macros_translator.cpp:77]: `throw` do it

File: MacrosTranslator/src/translator/commands/commonCommands/unique_commands.h (hunk 15 above)
Comment (smousy) [reply MacrosTranslator/src/translator/commands/commonCommands/unique_commands.h:263]: to optimize this code I will spend more time then this code will be used, this feature is for one time using, I think it don't need to be ideal optimize

File: MacrosTranslator/CMakeLists.txt
Comment (NikitaZotov) [inline MacrosTranslator/CMakeLists.txt:1]: Rename, please, MacrosTranslator to scn-macros-translator

**Review by NikitaZotov — APPROVED** [review]

**Review by VasKho — APPROVED** [review]

**Pull request merged**